packages/smart-contracts/src/contracts/BatchPayments.sol

```diff
@@ -0,0 +1,178 @@
+     * @param _tokenAddress Token to transact with.
+     * @param _recipients List of recipients accounts.
+     * @param _amounts List of amounts, corresponding to recipients[].
+     * @param _paymentReferences List of paymentRefs, corr. to the recipients[] and .
+     * @param _feeAmounts List of amounts of the payment fee, corr. to the recipients[].
+     * @param _feeAddress The fee recipient.
+     * @dev Uses ERC20FeeProxy.sol to pay an invoice and fees, with a payment reference.
+     */
+    function batchERC20PaymentsWithReferenceAndFee(
+        address _tokenAddress, 
+        address[] calldata _recipients, 
+        uint256[] calldata _amounts,
+        bytes[] calldata _paymentReferences,
+        uint256[] calldata _feeAmounts,
+        address _feeAddress 
+    ) external {
+        for (uint256 i = 0; i < _recipients.length; i++) {
+           (bool status, ) = address(paymentErc20FeeProxy).delegatecall(
+            abi.encodeWithSignature(
+            "transferFromWithReferenceAndFee(address,address,uint256,bytes,uint256,address)",
+                _tokenAddress,
+                _recipients[i], 
+                _amounts[i],
+                _paymentReferences[i],
+                _feeAmounts[i],
+                _feeAddress
+                )
+            );
+            require(status, "transferFromWithReference failed");
+        }
+    }
+
+    /**
+     * @notice Send a batch of erc20 payments on multiple tokens w/fees with paymentReferences to multiple accounts.
+     * @param _tokenAddresses List of tokens to transact with.
+     * @param _recipients List of recipients accounts.
+     * @param _amounts List of amounts, corresponding to recipients[].
+     * @param _paymentReferences List of paymentRefs, corr. to the recipients[].
+     * @param _feeAmounts List of amounts of the payment fee, corr. to the recipients[].
+     * @param _feeAddress The fee recipient.
+     * @dev Uses ERC20FeeProxy.sol to pay an invoice and fees, with a payment reference.
+     */
+    function batchERC20PaymentsMultiTokensWithReferenceAndFee(
+        address[] calldata _tokenAddresses, 
+        address[] calldata _recipients, 
+        uint256[] calldata _amounts,
+        bytes[] calldata _paymentReferences,
+        uint256[] calldata _feeAmounts,
+        address _feeAddress 
+    ) external {
+        for (uint256 i = 0; i < _recipients.length; i++) {
+           (bool status, ) = address(paymentErc20FeeProxy).delegatecall(
+            abi.encodeWithSignature(
+            "transferFromWithReferenceAndFee(address,address,uint256,bytes,uint256,address)",
+                _tokenAddresses[i],
+                _recipients[i], 
+                _amounts[i],
+                _paymentReferences[i],
+                _feeAmounts[i],
+                _feeAddress
+                )
+            );
+            require(status, "transferFromWithReference failed");
```

Comment:
does this mean one failure will fail the whole batch transaction? will it revert successful ones?

Reply:
We need to be sure of that choice.

What would be good for him about the "All or nothing" ?

One advantage of this approach is the gas estimation should fail if one of the payments is doomed to fail, which is better for UX in case of a bug.

One drawback is that if the failure is not predictable (eg the ETH balance got drained in a previous tx from the same block), the user may find it handy to have paid "as much as possible"

Reply:
I agree, and I need to take a decision early because it can modify quite a lot the code:

"nothing": 
pros ->
- if X txs failed, customer has "just" to send again the amount these X txs, even without batch
- if there is a failed, gas already used is not wasted
cons ->
- right now, batchEthereum revert all, and will need an important change to avoid it. (because it cannot use delegatecall with msg.valueSpecificForEachTx, so I have to use an interface of feeProxy, -> the transaction reverted will revert the batch directly. I only know one way to use a delegatecall but it will use a lot of gas)

"all":
pros ->
- "gas estimation should fail if one of the payments is doomed to fail" Yoann
cons:
- "if the failure is not predictable" Yoann, a lot of customer's gas can be wasted